book/chrome.md

```diff
@@ -25,91 +25,102 @@ will define `handle_click` and bind it to `<Button-1>`, button number
 
 
 ``` {.python}
-window.bind("<Button-1>", handle_click)
-
-def handle_click(e):
-    pass
+class Browser:
+    def __init__(self):
+        # ...
+        self.window.bind("<Button-1>", self.handle_click)
 ```
 
 Inside `handle_click`, we want to figure out what link the user has
 clicked on. We'll need to look at the `e` argument, which contains an
-\"event object\". This object has `x` and `y` fields, which refer to
+"event object". This object has `x` and `y` fields, which refer to
 where the click happened, relative to the corner of the browser
 window. Since the canvas is in the top-left corner of the window,
 those are also the *x* and *y* coordinates relative to the canvas. To
 get coordinates relative to the web page, we need to account for
 scrolling:
 
-``` {.python}
-x, y = e.x, e.y + scrolly
+``` {.python expected=False}
+def handle_click(self, e):
+    x, y = e.x, e.y + self.scroll
 ```
 
 The next step is to figure out what links or other elements are at
-that location. Naively, this seems like it should be easy. We already
-have a tree of layout objects, each of which records size and
-position. We could use those to find the element clicked on, something
-like this:
+that location. To do that, we need to keep the tree of layout objects
+(which right now we create and then throw away) available as a field
+on the `Browser` class:
 
 ``` {.python}
-def handle_click(e):
-    x, y = e.x, e.y + scrolly
-    elt = find_element(x, y, mode)
-    print(elt.tag)
+def layout(self, tree):
+    self.document = DocumentLayout(tree)
+    # ...
 ```
 
-Here the `find_element` function is a straightforward variant of code
+Now `handle_click` can walk this tree, checking each layout object's
+size and position to find the element clicked on:
+
+``` {.python}
+def handle_click(self, e):
+    # ...
+    obj = find_layout(x, y, self.document)
+    if not obj: return
+    elt = obj.node
+```
+
+Here the `find_layout` function is a straightforward variant of code
 we've already written a few times:
 
 ``` {.python}
-def find_element(x, y, layout):
-    for child in reversed(layout.children):
-        result = find_element(x, y, child)
+def find_layout(x, y, tree):
```

Comment:
why rename this?

Reply:
I think it used to return the element you clicked on, and now returns the layout object you clicked on. We need the layout object because reflow takes a layout object; not sure how it worked before.

Reply:
How about find_layout_object?

Reply:
That's probably better, but for code organization purposes I'd rather eliminate all of these `find_X` methods. Let's discuss next meeting how to get rid of them.